Drivers: ADC: Introduce gd32 adc driver

cameled opened the pull request:
Add an initial support for gd32 adc interface.

This driver based on single conversion mode with regular channel.

Test with `samples/drivers/adc` sample on gd32f350r_eval, gd32f450i_eval and gd32vf103v_eval boards.

work list:
 - [x] sample time config

cameled commented:
When compile the `samples/drivers/adc` sample on gd32f350r_eval board, I get an ` warning: unused variable 'cfg'` at adc_gd32.c. gd32f450i_eval and gd32vf103v_eval boards doesn't have this issue. 

I guess this issue caused by the ADC interface number. GD32F350 have one ADC interface, but GD32F450 and GD32VF103 have more than one ADC interface. And I use  the `DT_INST_FOREACH_STATUS_OKAY` macro.

Warning output: [warning.txt](https://github.com/zephyrproject-rtos/zephyr/files/8168292/warning.txt)?

gmarull reviewed (commented):
LGTM, a few minor comments. Boards `board.yaml` should also be updated to include `adc` on the supported list.

cameled commented on samples/drivers/adc/boards/gd32f350r_eval.overlay:
```diff
@@ -0,0 +1,12 @@
+
+ / {
+	zephyr,user {
+		/* adjust channel number according to pinmux in board.dts */
```
All overlay under `samples/drivers/adc` use the same comment. I want to keep this pattern in here.

cameled commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	uint16_t *repeat_buffer;
+};
+
+/* Use a flag to protect the shared irq handler initialization. */
+static bool init_irq = true;
```
shared_irq implement refer from #41497, thanks for @ycsin

ycsin replied:
> shared_irq implement refer from #41497, thanks for @ycsin

Credit to @mcscholtz

ycsin replied:
I actually still think that this variable can be local to `adc_gd32_shared_irq_configure`, instead of public

cameled replied:
```

cameled replied:
Move to `adc_gd32_shared_irq_configure`, `init_irq` used to protect the function just run at once.

cameled commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+		return ret;
+	}
+
+#ifdef CONFIG_SOC_SERIES_GD32F3X0
+	/* Select APB2 clock as ADC clock source. */
+	rcu_adc_clock_config(RCU_ADCCK_APB2_DIV4);
+#endif
```
Sure, I will add a property to hold it.

cameled replied:
done.

cameled commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (chan_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("gain configure not supported");
```
done.

cameled commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	}
+
+	if (chan_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("reference configure not supported");
```
done.

gmarull commented on boards/arm/gd32e103v_eval/doc/index.rst:
```diff
@@ -68,7 +68,7 @@ The board configuration supports the following hardware features:
      - N/A
    * - USART
-     - :kconfig:`CONFIG_SERIAL`
+     - :kconfig:option:`CONFIG_SERIAL`
```
nit: has already been fixed: https://github.com/zephyrproject-rtos/zephyr/commit/656beff50a9fd013c5f0458b688a9edb7f290eea

cameled replied:
removed.

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	const struct adc_gd32_config *cfg = dev->config;
+
+	if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
+		*data->buffer++ = ADC_RDATA(cfg->reg);
```
I could be wrong but I remember this is considered ambiguous according to one of the misra c's rule

cameled replied:
`*data->buffer++` is widely used in adc drivers.

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_gd32_data *data =
+		CONTAINER_OF(ctx, struct adc_gd32_data, ctx);
```
The column width has been increased to 100, I think you should be able to fit this into a single line

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+		return -EINVAL;
+	}
+
+	index = 0;
```
why not init this as you declare it?

cameled replied:
Just want to explicitly declared that we assign `0` to `index` to represent `ADC_ACQ_TIME_DEFAULT`.

Remove this assignment, merge to the `index` declaration.

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	if (acq_time != ADC_ACQ_TIME_DEFAULT) {
+		for (uint8_t i = 0; i < ARRAY_SIZE(acq_time_tbl); i++) {
+			if (ADC_ACQ_TIME_VALUE(acq_time) <= acq_time_tbl[i]) {
+				index = i;
```
Seems like you should be able to use index as the incrementor instead of creating another `uint8_t i`?

cameled replied:
done.

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	}
+
+	if (chan_cfg->differential) {
+		LOG_ERR("differential sampling not supported");
```
If you factor these (and other similar ones) to:

```c
LOG_ERR("%s not supported", "differential sampling");
```

you would be able to save quite a few bytes in ROM, since `%s not supported` is shared.

cameled replied:
I think it's OK, for a better read experience reason..

cameled replied:
I adjust the log message.
`gain` and `reference` will not give "not supported" message to user, instead with `invalid` message.

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+
+	index = find_lsb_set(sequence->channels) - 1;
+	if (sequence->channels > BIT(index)) {
+		LOG_ERR("Only single channel supported");
```
```suggestion
		LOG_ERR("Only %s supported", "single channel");
```

ycsin commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,396 @@
+	}
+
+	if (chan_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Only ADC internal reference supported");
```
```suggestion
		LOG_ERR("Only %s supported", "ADC internal reference");
```

cameled commented:
> When compile the `samples/drivers/adc` sample on gd32f350r_eval board, I get an ` warning: unused variable 'cfg'` at adc_gd32.c. gd32f450i_eval and gd32vf103v_eval boards doesn't have this issue.
> 
> I guess this issue caused by the ADC interface number. GD32F350 have one ADC interface, but GD32F450 and GD32VF103 have more than one ADC interface. And I use the `DT_INST_FOREACH_STATUS_OKAY` macro.
> 
> Warning output: [warning.txt](https://github.com/zephyrproject-rtos/zephyr/files/8168292/warning.txt)

Caused by the macro redefinitions. For example, `#define ADC_CTL0(_) REG32(ADC + 0x00000004U)` doesn't use the `_`. so, cfg will be treated as unused variable.

nandojve commented on dts/bindings/adc/gd,gd32-adc.yaml:
```diff
@@ -0,0 +1,54 @@
+      required: false
+      default: 4
+      enum:
+        - 0 # RCU_ADCCK_IRC28M_DIV
+        - 1 # RCU_ADCCK_IRC28M
+        - 2 # RCU_ADCCK_APB2_DIV2
+        - 3 # RCU_ADCCK_AHB_DIV3
+        - 4 # RCU_ADCCK_APB2_DIV4
+        - 5 # RCU_ADCCK_AHB_DIV5
+        - 6 # RCU_ADCCK_APB2_DIV6
+        - 7 # RCU_ADCCK_AHB_DIV7
+        - 8 # RCU_ADCCK_APB2_DIV8
+        - 9 # RCU_ADCCK_AHB_DIV9
```
Is this range 0-9 valid for all gd32 SoCs?
if yes, all comments are valid for all gd32 SoCs? I mean, value 1 always be IRC28M? Can any other enum position assume other value?

cameled replied:
Refer from https://github.com/zephyrproject-rtos/zephyr/pull/41855#issuecomment-1019416536, this property currently only valid for gd32f3x0 board.

rcu adc clock enum definition can be found in [rcu_adc_clock_enum](https://github.com/zephyrproject-rtos/hal_gigadevice/blob/cc85acbd635f02122c54df11fa55458269ce1bdd/gd32f3x0/standard_peripheral/include/gd32f3x0_rcu.h#L441)

cameled replied:
Update the property description, support all external clock source.

nandojve commented on samples/drivers/adc/sample.yaml:
```diff
@@ -5,7 +5,8 @@ tests:
     depends_on: adc
     platform_allow: nucleo_l073rz disco_l475_iot1 cc3220sf_launchxl
-        cc3235sf_launchxl stm32l496g_disco
+        cc3235sf_launchxl stm32l496g_disco gd32f350r_eval gd32f450i_eval
+        gd32vf103v_eval
```
Why not enable F403? Is there any particular reason?

cameled replied:
No GD32F403 evaluation board on my side.

adc-gd32 is a general GD32 ADC driver. So, I think it works on GD32F403 too.

cameled replied:
There have two remain issue for gd32f403 ADC.
1. pinconfig for GD32F403Zx in hal_gigadevice missing PC3. (To run the ADC sample on GD32F403Z-EVAL board)
2. ADC0 and ADC1 shared interrupt 18, but ADC2 use an individual interrupt 47.

So, after fix the PC3 missing issue, ADC0 and ADC1 will work well. But ADC2 looks need more work to adjust adc_gd32 driver.

cameled replied:
gd32f403z_eval also support the sample `samples/drivers/adc`

Commit:
dts: introduce gd32 adc

Add support for gd32 adc.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
drivers: adc: introduce gd32 adc driver

This driver based on single conversion mode with regular channel.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
boards: add adc support for gd32 boards

This add adc support for gd32f350r_eval, gd32f450i_eval and
gd32vf103v_eval boards.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

Commit:
samples: drivers: adc: add gd32 boards support

This add gd32 boards support for the adc sample.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

cameled commented:
@gmarull @nandojve Could you please help to take a look at this? Thanks!

Commit:
MAINTAINERS: add cameled as gd32 platform collaborator

Add cameled as gd32 platform collaborator.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

gmarull commented on dts/bindings/adc/gd,gd32-adc.yaml:
```diff
@@ -0,0 +1,70 @@
+        RCU(Reset Control Unit) HAL provide a enum called rcu_adc_clock_enum to
+        select the ADC clock source.
+
+        This use an integer to mapping the rcu_adc_clock_enum, and may have
+        different meaning between GD32 SoCs. For example, GD32F3X0 RCU
+        rcu_adc_clock_enum mapping as below:
+          - 0 RCU_ADCCK_IRC28M_DIV2
+          - 1 RCU_ADCCK_IRC28M
+          - 2 RCU_ADCCK_APB2_DIV2
+          - 3 RCU_ADCCK_AHB_DIV3
+          - 4 RCU_ADCCK_APB2_DIV4
+          - 5 RCU_ADCCK_AHB_DIV5
+          - 6 RCU_ADCCK_APB2_DIV6
+          - 7 RCU_ADCCK_AHB_DIV7
+          - 8 RCU_ADCCK_APB2_DIV8
+          - 9 RCU_ADCCK_AHB_DIV9
+      enum:
+        - 0
+        - 1
+        - 2
+        - 3
+        - 4
+        - 5
+        - 6
+        - 7
+        - 8
+        - 9
```
better provide definitions in `include/dts-bindings/`, see this out of tree example: https://github.com/teslabs/spinner/blob/main/include/dts-bindings/adc/stm32f3xx.h.

cameled replied:
Thanks for you advice!

cameled replied:
I add a new commit to hold this. If everything is OK, it will be merge to old commit.

gmarull commented on dts/bindings/adc/gd,gd32-adc.yaml:
```diff
@@ -24,31 +24,6 @@ properties:
         RCU(Reset Control Unit) HAL provide a enum called rcu_adc_clock_enum to
         select the ADC clock source.
```
nit: add comment that valid values are in `dts-bindings/adc/gd32xxx.h` headers.

cameled replied:
done.

gmarull commented on include/dt-bindings/adc/gd32f3x0.h:
```diff
@@ -0,0 +1,20 @@
+#ifndef ZEPHYR_INCLUDE_DT_BINDINGS_ADC_GD32F3X0_H_
+#define ZEPHYR_INCLUDE_DT_BINDINGS_ADC_GD32F3X0_H_
+
+#define RCU_ADCCK_IRC28M_DIV2	0
```
I'd add `GD32_` prefix here.

cameled replied:
Just want to give a redefinition for the rcu_adc_clock_enum in here.

OK, I will change it.

cameled replied:
done.

nandojve commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,403 @@
+	return 0;
+}
+
+#define HANDLE_IRQS(idx)							\
+	static const struct device *dev_##idx = DEVICE_DT_INST_GET(idx);	\
+	const struct adc_gd32_config *cfg_##idx = dev_##idx->config;		\
+										\
+	if (ADC_CTL0(cfg_##idx->reg) & ADC_CTL0_EOCIE) {			\
+		adc_gd32_isr(dev_##idx);					\
+	}
+
+static void adc_gd32_shared_irq_handler(void)
+{
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_gd32_shared_irq_configure(void)
+{
+	/* Use a flag to protect the shared irq handler initialization. */
+	static bool init_irq = true;
+
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    adc_gd32_shared_irq_handler,
+			    NULL, 0);
+		irq_enable(DT_INST_IRQN(0));
+	}
+}
```
The same way to use to create `adc_gd32_config_##n` you can add the IRQ handle. This way you can have only 1 test each time IRQ is activated when have multiple instances.

cameled replied:
I'm trying to implement it, but failed. Any detailed explanation? Thanks!

nandojve replied:
I was wondering if something like below could work:

```diff
diff --git a/drivers/adc/adc_gd32.c b/drivers/adc/adc_gd32.c
index 3d45df6acc..5590bc6c04 100644
--- a/drivers/adc/adc_gd32.c
+++ b/drivers/adc/adc_gd32.c
@@ -92,6 +92,10 @@ static void adc_gd32_isr(const struct device *dev)
        struct adc_gd32_data *data = dev->data;
        const struct adc_gd32_config *cfg = dev->config;
 
+       if (!(ADC_CTL0(cfg->reg) & ADC_CTL0_EOCIE)) {
+               return;
+       }
+
        if (ADC_STAT(cfg->reg) & ADC_STAT_EOC) {
                *data->buffer++ = ADC_RDATA(cfg->reg);
 
@@ -344,34 +348,6 @@ static int adc_gd32_init(const struct device *dev)
        return 0;
 }
 
-#define HANDLE_IRQS(idx)                                                       \
-       static const struct device *dev_##idx = DEVICE_DT_INST_GET(idx);        \
-       const struct adc_gd32_config *cfg_##idx = dev_##idx->config;            \
-                                                                               \
-       if (ADC_CTL0(cfg_##idx->reg) & ADC_CTL0_EOCIE) {                        \
-               adc_gd32_isr(dev_##idx);                                        \
-       }
-
-static void adc_gd32_shared_irq_handler(void)
-{
-       DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
-}
-
-static void adc_gd32_shared_irq_configure(void)
-{
-       /* Use a flag to protect the shared irq handler initialization. */
-       static bool init_irq = true;
-
-       if (init_irq) {
-               init_irq = false;
-               IRQ_CONNECT(DT_INST_IRQN(0),
-                           DT_INST_IRQ(0, priority),
-                           adc_gd32_shared_irq_handler,
-                           NULL, 0);
-               irq_enable(DT_INST_IRQN(0));
-       }
-}
-
 #ifdef CONFIG_SOC_SERIES_GD32F3X0
 #define ADC_CLOCK_SOURCE(n)                                                    \
        .rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock)
@@ -381,6 +357,15 @@ static void adc_gd32_shared_irq_configure(void)
 
 #define ADC_GD32_INIT(n)                                                       \
        PINCTRL_DT_INST_DEFINE(n);                                              \
+       static void adc##n##_gd32_irq_config(const struct device *dev)          \
+       {                                                                       \
+               IRQ_CONNECT(DT_INST_IRQN(n),                                    \
+                           DT_INST_IRQ(n, priority),                           \
+                           adc_gd32_isr,                                       \
+                           DEVICE_DT_INST_GET(n),                              \
+                           0);                                                 \
+               irq_enable(DT_INST_IRQN(n));                                    \
+       }                                                                       \
        static struct adc_gd32_data adc_gd32_data_##n = {                       \
                ADC_CONTEXT_INIT_TIMER(adc_gd32_data_##n, ctx),                 \
                ADC_CONTEXT_INIT_LOCK(adc_gd32_data_##n, ctx),                  \
@@ -391,7 +376,7 @@ static void adc_gd32_shared_irq_configure(void)
                .rcu_periph_clock = DT_INST_PROP(n, rcu_periph_clock),          \
                .channels = DT_INST_PROP(n, channels),                          \
                .pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),                      \
-               .irq_config_func = adc_gd32_shared_irq_configure,               \
+               .irq_config_func = adc##n##_gd32_irq_config,                    \
                ADC_CLOCK_SOURCE(n)                                             \
        };                                                                      \
        DEVICE_DT_INST_DEFINE(n,                                                \
```

cameled replied:
Normally, GD32 ADC sharing the interrupt between each other.

The main problem is how to get ADC0...ADCn device pointers. Without device pointer, we can not dispatch interrupt to each GD32 ADC.

cameled replied:
Refer from [Zephyr interrupt document](https://docs.zephyrproject.org/latest/reference/kernel/other/interrupts.html).
> Only a single ISR can be associated with a specific IRQ at any given time.

cameled commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,403 @@
+{
+	uint8_t index = 0, offset;
+
+	/* Acquisitioin time unit is adc clock cycle. */
```
done.

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,403 @@
+	DT_INST_FOREACH_STATUS_OKAY(HANDLE_IRQS);
+}
+
+static void adc_gd32_shared_irq_configure(void)
+{
+	/* Use a flag to protect the shared irq handler initialization. */
+	static bool init_irq = true;
+
+	if (init_irq) {
+		init_irq = false;
+		IRQ_CONNECT(DT_INST_IRQN(0),
+			    DT_INST_IRQ(0, priority),
+			    adc_gd32_shared_irq_handler,
+			    NULL, 0);
+		irq_enable(DT_INST_IRQN(0));
```
this assume information from instance 0 is always the correct one, is that always the case?

cameled replied:
Correct for most of gd32 soc.

As I know there have a exception for gd32f403.  gd32f403 have 3 ADC. ADC0 and ADC1 sharing the interrupt 18, but ADC3 use an individual interrupt 47.

I have tried several way to fix gd32f403 ADC2. The main goal is fix it at preprocess times:
  - use DTS macro, like DT_FOREACH;
  - use DTS property, like `sharing-interrupts`.

Finally, I'm failed. So I just ignore the ADC2, gd32f403 DTS only have ADC0 and ADC2 now.

ycsin replied:
The irq sharing will only work when all of the ADC share the same interrupt line.
You might be able to workaround this with hardcoded `#ifdef` if there is not a lot of combinations, otherwise you will have the same issue as #39565 which no one has a fix yet.

cameled replied:
After looks all gd32 series user manual, I get below ADC interrupt table.

|    GD32 SoC    |    ADC0 IRQN    |    ADC1 IRQN    |    ADC2 IRQN    |
|     -----------       |       -----------        |    -----------          |       -----------        |
|  GD32F3x0      |
|  GD32E23x      |
|  GD32E23x      |

ycsin commented on dts/bindings/adc/gd,gd32-adc.yaml:
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2021 BrainCo Inc.
```
Probably you can update this (and others that apply) to 2022.

nandojve commented:
Hi @cameled ,

Is there any update on this?

cameled commented:
> Hi @cameled ,
> 
> Is there any update on this?

Sorry for the long delay, I switch to another task last month. Fortunately, Its almost finish now.
I can restart this work at the end of this week.

cameled commented:
@ycsin @nandojve @gmarull I use a runtime solution with `adc_gd32_global_irq_handler` to fix the irregular interrupt implement. Cloud you please have a look for this?

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,435 @@
+#define DT_DRV_COMPAT gd_gd32_adc
+
+#include <errno.h>
+#include <drivers/pinctrl.h>
+#include <drivers/adc.h>
```
nit: please use `<zephyr/>` prefix

cameled replied:
thanks! done.

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,435 @@
+};
+#endif
+
+static uint8_t adc_gd32_init_irqs[8U];
```
note: why is this array size set to 8? can this be explained?

cameled replied:
Use `GD32_ADC_NUM_MAX` macro instread. GD32 ADC number always smaller than 3.

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,435 @@
+#define ADC_CLOCK_SOURCE(n)
+#endif
+
+#define DT_ADC2 0U
```
not used?

cameled replied:
removed.

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,435 @@
+		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(n),					\
+		.irq_num = DT_INST_IRQN(n),							\
+		.irq_config_func = adc_gd32_irq_cfg_##n,					\
+		ADC_CLOCK_SOURCE(n)								\
```
why not:
```
IF_ENABLED(CONFIG_SOC_SERIES_GD32F3X0, (.rcu_clock_source = DT_INST_PROP(n, rcu_periph_clock),))
```

cameled replied:
There also have others gd32 soc contain additional adc clock.

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,435 @@
+
+#define ADC_GD32_INIT(n)									\
+	PINCTRL_DT_INST_DEFINE(n);								\
+	static void adc_gd32_irq_cfg_##n(void)							\
+	{											\
+		bool initialized = false;							\
+												\
+		for (size_t i = 0; i < sizeof(adc_gd32_init_irqs); i++) {			\
+			if (adc_gd32_init_irqs[i] == 0U) {					\
+				initialized = true;						\
+				adc_gd32_init_irqs[i] = DT_INST_IRQN(n);			\
+				break;								\
+			}									\
+												\
+			if (adc_gd32_init_irqs[i] == DT_INST_IRQN(n)) {				\
+				LOG_INF("interrupt initialize success, num: %u",		\
+					DT_INST_IRQN(n));					\
+				return;								\
+			}									\
+		}										\
+												\
+		if (initialized) {								\
+			IRQ_CONNECT(DT_INST_IRQN(n),						\
+				DT_INST_IRQ(n, priority),					\
+				adc_gd32_global_irq_handler,					\
+				DEVICE_DT_INST_GET(n),						\
+				0);								\
+			irq_enable(DT_INST_IRQN(n));						\
+			LOG_INF("interrupt initialize success, num: %u, priority: %u",		\
+				DT_INST_IRQN(n), DT_INST_IRQ(n, priority));			\
+		} else {									\
+			LOG_INF("interrupt initialize failed");					\
+		}										\
+	}											\
```
I'd personally do the following instead:

1. create a "connect" function that just has `IRQ_CONNECT()` for every instance
2. Move all the initialization logic to the init/sub-init function. If an IRQ hasn't been initialized, just run the "connect" call for that instance. 

Maybe I'm missing something, but I think it's doable.

cameled replied:
Looks `IRQ_CONNECT` have a `build time` restriction, refer from [aarch32 irq definition](https://github.com/zephyrproject-rtos/zephyr/blob/2ef3fe7bca137dcbf6bad3b13a038d1998aa1a37/include/zephyr/arch/arm/aarch32/irq.h#L115).

cameled replied:
So, this solution still have problem. It should provide an unique `IRQ_CONNECT` for each interrupt.

cameled replied:
Config gd32 adc irq through node-label, based on https://github.com/zephyrproject-rtos/zephyr/pull/42215#discussion_r873123172

gmarull reviewed (commented):
Thanks for this contribution! Please check my comments

Commit:
dst: gigadevice: change usart3 address to lower case

DTS node address should use lower case. This fix an incorrect
gd32f403 usart3 node address.

Signed-off-by: HaiLong Yang <hailong.yang@brainco.cn>

gmarull commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,462 @@
+#undef ADC_RSQ2
+#undef ADC_RDATA
+
+#define ADC0			ADC
```
this definition is not used

cameled replied:
removed.

cameled commented:
@nandojve Please help to confirm the new shared irq implement, thanks!

nandojve commented on drivers/adc/adc_gd32.c:
```diff
@@ -0,0 +1,461 @@
+
+static void adc_gd32_global_irq_cfg(void)
+{
+	static bool global_irq_init = true;
+
+	if (!global_irq_init) {
+		return;
+	}
+
+	global_irq_init = false;
```
Why not use positive logic?

```c
	static bool global_irq_init = false;

	if (global_irq_init) {
		return;
	}

	global_irq_init = true;
```

cameled replied:
done.

cameled replied:
An error occured, `ERROR:INITIALISED_STATIC: do not initialise statics to false`.
Revert to positive init value.

nandojve commented on boards/arm/gd32f403z_eval/gd32f403z_eval.dts:
```diff
@@ -88,6 +88,12 @@
 	status = "okay";
 };
 
+&adc0 {
```
Missing entry at https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/gd32f403z_eval/doc/index.rst#supported-features

cameled replied:
done.

nandojve reviewed (changes requested):
Hi @cameled,

I found more two small fixes. Overall is a very nice work!
